### PR TITLE
Localize resources

### PR DESCRIPTION
**Pull request by annasong20**
This PR localizes entries under the `resources` kustomization field.

### CODE REVIEW

File: api/internal/localizer/errors.go
Comment (yuwenma) [inline api/internal/localizer/errors.go:25]: super nit. remove this line

**Review by yuwenma — CHANGES_REQUESTED** [review]

File: api/internal/localizer/errors.go
Comment (annasong20) [reply api/internal/localizer/errors.go:25]: @natasha41575 recommended this for readability: https://github.com/kubernetes-sigs/kustomize/pull/4869#discussion_r1033740486

I'm down either way.

File: api/internal/localizer/errors.go
Comment (annasong20) [reply api/internal/localizer/errors.go:28]: @natasha41575 recommended this for `ResourceLoadError`: https://github.com/kubernetes-sigs/kustomize/pull/4869#discussion_r1024252376

I applied it here. The reasoning is that `localizeFile` could've failed not because the path isn't a file, but, say, because the file fails `fileLoader` or localize constraints. In this case, the user will be confused by the `localizeDir` error message because they know that path is a file.

An alternative is exposing a public error in `fileLoader` when the `Load` path is not a file. Then, we can check for these errors at the localize level and only try `localizeDir` when `localizeFile` returns such an error. There is actually already an issue for this: #4807

File: api/internal/localizer/errors.go
Comment (natasha41575) [reply api/internal/localizer/errors.go:25]: I think for errors, lines like this are not necessary.

File: api/internal/localizer/errors.go
Comment (natasha41575) [reply api/internal/localizer/errors.go:28]: > The reasoning is that localizeFile could've failed not because the path isn't a file, but, say, because the file fails fileLoader or localize constraints. In this case, the user will be confused by the localizeDir error message because they know that path is a file.

Yeah, I've run into a couple of issues in the past with similar code like "if it's not a file, try loading it as something else", and have gotten issues filed by users when the error didn't make sense in some cases. For that reason, now I prefer to have both errors exposed. I don't think it necessarily needs to be multiple lines; semicolons might be better than newlines to make it look more like a typical error. Maybe something more like `could not load path %q as file: %s; could not load path %q as directory: %s` would look cleaner.

File: api/internal/localizer/localizer_test.go
Comment (natasha41575) [inline api/internal/localizer/localizer_test.go:757]: Could we add a tests here with a remote resource file, and a remote resource directory?

Comment (annasong20) [reply api/internal/localizer/localizer_test.go:757]: We can't add that test yet because the code that calculates the path to save a remote root, #4909, hasn't been merged yet.

Also, since these are integration tests, I planned to add them in a separate PR in `krusty`. I can add them in `krusty` in this PR once #4909 is merged?

Comment (natasha41575) [reply api/internal/localizer/localizer_test.go:757]: I thought that the `localizeFile` and `localizeDir` functions are supposed to be converting the remote URLs to local filepaths, so IMO these tests should cover that behavior. Can we get #4909 in and then add the tests here?

File: api/internal/localizer/errors.go
Comment (yuwenma) [reply api/internal/localizer/errors.go:25]: Yeah, in general the `var _ InterfaceX = StructX{}` is helpful since golang does not require specifying the implemented interfaces, so this static check makes the code author's purpose more clear. 
Though the `Error` interface is much less likely to change and need this sanity check. It's more like a convention.

File: api/internal/localizer/errors.go
Comment (yuwenma) [reply api/internal/localizer/errors.go:28]: I see the reasoning here. So basically you want to expose all the assumptions and corresponding logics and let users triage the issues.  I'm ok with that.
Defer to @annasong20

**Review by yuwenma — APPROVED** [review]
LGTM on my side. Can you resolve @natasha41575 comment about the unittest before merging?

File: api/internal/localizer/localizer_test.go
Comment (natasha41575) [reply api/internal/localizer/localizer_test.go:757]: I'm also OK to merge this before #4909, if we add the tests in #4909. Let me know which you prefer.

Comment (annasong20) [reply api/internal/localizer/localizer_test.go:757]: > I thought that the `localizeFile` and `localizeDir` functions are supposed to be converting the remote URLs to local filepaths, so IMO these tests should cover that behavior. Can we get #4909 in and then add the tests here?

Sure, sg.

File: api/internal/localizer/errors.go
Comment (annasong20) [reply api/internal/localizer/errors.go:25]: Done.

File: api/internal/localizer/errors.go
Comment (annasong20) [reply api/internal/localizer/errors.go:28]: Done.

File: api/krusty/localizer/runner_test.go
Comment (natasha41575) [inline api/krusty/localizer/runner_test.go:204]: Since we have a lot of helper functions now, can we move these helpers to their own file `api/krusty/localizer/testutils.go`? I think it will help keep this directory organized.

**Review by natasha41575 — COMMENTED** [review]
Tests LGTM. I just have a small comment about organization.

File: api/krusty/localizer/runner_test.go
Comment (annasong20) [reply api/krusty/localizer/runner_test.go:204]: I moved all the helper functions to the top of the file for now. I'm all for improving organization. The problem with moving the helper functions to `testutils.go` is this file can't exist in package `localizer_test` because the file name doesn't end in `_test.go`. We don't want the test helper functions to be publicly exposed in the `localizer` package either. 

If moving the helper functions to the top of the file is not clean enough, I came up with a few other options:
1. Create a `loctest` package under `api/testutils` and move the helper functions to a file there called `helper.go`.
2. Create a `testutils` package under `api/krusty/localizer/internal` and move the helper functions to a file there called `helper.go`.
3. Move the helper functions to a new `utils_test.go` file under `api/krusty/localizer`.

**Review by natasha41575 — APPROVED** [review]
/lgtm

**Pull request merged**
